components/org.wso2.carbon.uis/src/main/java/org/wso2/carbon/uis/internal/io/deployment/ArtifactAppDeployer.java

```diff
@@ -27,125 +27,170 @@
-                    .map(ArtifactAppReference::new)
-                    .collect(Collectors.toSet());
-        } catch (IOException e) {
-            throw new DeploymentException("Cannot list web apps in '" + appsRepository + "' directory.", e);
+    @Override
+    public Object deploy(Artifact artifact) throws CarbonDeploymentException {
+        Path appPath = artifact.getFile().toPath();
+        if (!isValidAppArtifact(appPath)) {
+            throw new DeploymentException("Artifact located in '" + appPath + "'is not a valid web app.");
         }
-        publishAppsDeploymentEvents(appReferences);
+
+        App app = createApp(appPath);
+        String artifactKey = "webapp:" + app.getName();
+
+        appDeploymentEventListener.appDeploymentEvent(app);
+        deployedApps.put(artifactKey, app);
+        LOGGER.debug("Web app '{}' deployed for context path '{}'.", app.getName(), app.getContextPath());
+        return artifactKey;
     }
 
-    private void stop() {
-        for (String appName : deployedAppNames) {
-            publishAppUndeploymentEvent(appName);
+    @Override
+    public void undeploy(Object key) throws CarbonDeploymentException {
+        App app = deployedApps.remove(key.toString());
+        if (app == null) {
+            throw new CarbonDeploymentException("Web app for key '{}' cannot be found to undeploy.");
         }
+
+        appDeploymentEventListener.appUndeploymentEvent(app.getName());
+        LOGGER.debug("Web app '{}' undeployed from context path '{}'.", app.getName(), app.getContextPath());
     }
 
-    private void publishAppDeploymentEvent(AppReference deployingAppReference) {
-        for (AppDeploymentEventListener appDeploymentEventListener : eventListeners) {
-            appDeploymentEventListener.appDeploymentEvent(deployingAppReference);
-        }
-        deployedAppNames.add(deployingAppReference.getName());
+    @Override
+    public Object update(Artifact artifact) throws CarbonDeploymentException {
+        // when an app artifact gets updated, we just re-deploy it
+        undeploy(artifact.getKey());
+        return deploy(artifact);
     }
 
-    private void publishAppsDeploymentEvents(Set<AppReference> deployingAppsReferences) {
-        eventListeners.forEach(listener -> listener.appsDeploymentEvents(deployingAppsReferences));
+    @Override
+    public URL getLocation() {
+        return deploymentLocation;
     }
 
-    private void publishAppUndeploymentEvent(String undeployingAppName) {
-        for (AppDeploymentEventListener appDeploymentEventListener : eventListeners) {
-            appDeploymentEventListener.appUndeploymentEvent(undeployingAppName);
+    @Override
+    public ArtifactType getArtifactType() {
+        return artifactType;
+    }
+
+    private boolean isValidAppArtifact(Path appPath) throws CarbonDeploymentException {
+        try {
+            return Files.exists(appPath) && Files.isDirectory(appPath) && Files.isReadable(appPath) &&
```

Comment:
Are we expecting a directory?

Reply:
Yes. A web UI app is contained in a directory. See test web app artifact [here](https://github.com/wso2/carbon-ui-server/tree/v0.9.3/tests/distribution/carbon-home/deployment/reactapps/test).

Reply:
Ack.